README.md
## Note ##

* This is forked from [crazy-max/docker-dokuwiki](https://github.com/crazy-max/docker-dokuwiki)
* Backup to Git repo is based on [ericbarch/dockuwiki](https://github.com/ericbarch/dockuwiki)

## About ##

* It is [Docker image](https://hub.docker.com/r/mtilson/dokuwiki/) of [DokuWiki](https://www.dokuwiki.org/dokuwiki) behind [Traefik](https://github.com/containous/traefik)
* It uses [Traefik Docker image](https://github.com/containous/traefik-library-image) 
* Traefik is used as reverse proxy and for unattended creation/renewal of Let's Encrypt certificates

## Changes on 2020/01/15 ##

* Used before this change the common `docker-compose.yml` file is separated now to one for 'docuwiki' application container (`docker-compose.yml`) and dedicated one for 'traefik' container (`traefik/docker-compose.yml`)
* Due to this change you can now run 'traefik' container independently from 'dokuwiki' application container, which can be useful in case you already have 'traefik' container running as an edge proxy for other application containers
  * The only common configuration item which needs to be shared between 'traefik' container and its served application containers is the name of their common network defined by `COMMON_NETWORK` variable, see below
* The given 'traefik' `docker-compose.yml` file (`traefik/docker-compose.yml`) can be used as an example, in case you'd like to run 'dokuwiki' application container on fresh docker system, as described below

## Features ##

* Alpine Linux 3.9, Nginx, PHP 7.2, ACME Let's Encrypt via Traefik
* Tarball authenticity checked during building process
* OPCache enabled to store precompiled script bytecode in shared memory
* Data, configuration, plugins, and templates are backed up to a configured Git repo

## Environment variables ##

* Variables defined in `.env` file
    * `DOKUWIKI_TRAEFIK_FE_RULE`
        * `traefik` frontend rule
        * set default to `Host:wiki.example.com` in `docker-compose.yml`
        * used in `docker-compose.yml`, if not defined traffic will not be routed to `dokuwiki` container
        * example
            * `DOKUWIKI_TRAEFIK_FE_RULE=Host:wiki.example.com`
    * `PERSISTENT_DIR`
        * host persistent volume to store DockuWiki site data, ACME Let's Encrypt certificates, and a pravite key of your Git backup server account
        * set default to `/opt/docker/persistent` in `pre-deploy.sh`
        * set default to `/opt/docker/persistent` in `docker-compose.yml`
        * used by `pre-deploy.sh` to create host persistent volumes directory structure
        * used in `docker-compose.yml` and `traefik/docker-compose.yml` to define persistent volumes
        * example
            * `PERSISTENT_DIR=/opt/docker/persistent`
    * `ACME_EMAIL`
        * email address used for ACME (Let's Encrypt) registration
        * set default to `webmaster@example.com` in `traefik/docker-compose.yml`
        * used in `traefik/docker-compose.yml` to define `acme` email address, if not defined Let's Encrypt will not work correctly
        * example
            * `ACME_EMAIL=webmaster@example.com`
    * `DOCKER_DOMAIN`
        * default base domain name used for the frontend rules
        * set default to `docker.localhost` in `traefik/docker-compose.yml`
        * used in `traefik/docker-compose.yml` to define base domain name for frontend rules for hosts which are not full domain name
        * example
            * `DOCKER_DOMAIN=docker.localhost`
    * `COMMON_NETWORK`
        * name of the network common for `traefik` and its served containers
        * set default to `traefik-public-network` in `docker-compose.yml`
        * set default to `traefik-public-network` in `traefik/docker-compose.yml`
        * used in `docker-compose.yml` and `traefik/docker-compose.yml` to define the name of the docker bridged network for connectivity
        * example
            * `COMMON_NETWORK=traefik-public-network`
    * `BACKUP_USER_EMAIL`
        * backup user email address
        * used to mark generated public key to be added to the account used to access Git backup repo
        * used to configure Git global option `user.email` and derive `user.name` (as the part of the email address before '@' sign) for Git commands used to commit backup data to Git backup repo
        * used in container ENTRYPOINT
        * passed from `docker-compose` to container ENTRYPOINT
        * set default to `dokuwiki-backup@example.com` in `docker-compose.yml`
        * set default to `dokuwiki-backup@example.com` in container ENTRYPOINT if passed empty
        * example
            * `BACKUP_USER_EMAIL=dokuwiki-backup@example.com`
    * `GIT_BACKUP_REPO_URL`
        * [Git remote URL](https://help.github.com/en/articles/about-remote-repositories) of your repo on Git server to backup wiki content
            * Git associates a remote URL with a name, which is called `origin` by default, and for which you can get the URL with the following command (run within your repo directory)
                * `git remote get-url origin`
        * mandatory and validated in `pre-deploy.sh` script and in container ENTRYPOINT
        * passed by `docker-compose` to container ENTRYPOINT
        * example
            * `GIT_BACKUP_REPO_URL=git@bitbucket.org:username/reponame.git`
    * `TZ`
        * container timezone
        * used in container ENTRYPOINT
        * set default to `Europe/Luxembourg` in `docker-compose.yml`
        * passed from `docker-compose` to container ENTRYPOINT
        * set default to `UTC` in container ENTRYPOINT if passed empty
        * example
            * `TZ=Europe/Oslo`
* The following variable have defaults and used in container ENTRYPOINT. You can redefine them in `.env` and they will be passed as is to container ENTRYPOINT by `docker-compose`
    * `MEMORY_LIMIT`
        * PHP memory limit
        * default to `256M`
        * example
            * `MEMORY_LIMIT=256M`
    * `UPLOAD_MAX_SIZE`
        * Upload max size
        * default to `16M`
        * example
            * `UPLOAD_MAX_SIZE=16M`
    * `OPCACHE_MEM_SIZE`
        * PHP OpCache memory consumption
        * default to `128`
        * example
            * `OPCACHE_MEM_SIZE=128`

## Volumes ##

* DokuWiki
    * `/data` - bind to host `${PERSISTENT_DIR}/dokuwiki/data` folder
        * folder that contains configuration, plugins, templates and data
    * `/root/.ssh` - bind to host `${PERSISTENT_DIR}/dokuwiki/root/.ssh` folder
        * folder that contains public/private keys, `config` and `known_hosts` files
        * you can place here the pravite key corresponding to a public key of your Git backup server account, name the file as `id_rsa`
* Traefik
    * `/acme.json` - bind to host `${PERSISTENT_DIR}/acme.json` file
        * file that contains ACME Let's Encrypt certificates

## Ports ##

* Traefik
    * `80` - HTTP port - redirects traffic to itself (Traefik) to HTTPS port (443)
    * `443` - HTTPS port - proxies traffic to DokuWiki to HTTP port (80)
* DokuWiki
    * `80` - HTTP port - serves DokuWiki wiki

## Installation ##

* On the fresh docker system (if you didn't run this installation procedure before), follow these steps:
    * Create project directory, `cd` to it, and run the following commands from within this project directory
    * Create `.env` file with the following environment variables, see the description and examples above
```bash
DOKUWIKI_TRAEFIK_FE_RULE=Host:wiki.example.com
PERSISTENT_DIR=/opt/docker/persistent
ACME_EMAIL=webmaster@example.com
DOCKER_DOMAIN=docker.localhost
BACKUP_USER_EMAIL=dokuwiki-backup@example.com
GIT_BACKUP_REPO_URL=git@bitbucket.org:username/reponame.git
COMMON_NETWORK=traefik-public-network
TZ=Europe/Oslo
MEMORY_LIMIT=
UPLOAD_MAX_SIZE=
OPCACHE_MEM_SIZE=
```
    * Download the pre-deployment script (`pre-deploy.sh`), make it executable, and run it
```bash
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/pre-deploy.sh > pre-deploy.sh
chmod +x pre-deploy.sh
./pre-deploy.sh
```
    * Provide access to your Git backup repo via SSH. You can do it the following way
        * Generate a public/pravite key pair
        * Place the private key to the host persistent volume as `${PERSISTENT_DIR}/dokuwiki/root/.ssh/id_rsa`
            * Make its permissions to be `read/write` only by `root`: 
                * `sudo chmod 600 ${PERSISTENT_DIR}/dokuwiki/root/.ssh/id_rsa`
        * Add the public key to the Git user account which has access to Git backup repo
            * `GIT_BACKUP_REPO_URL` variable defined above specifies *Git remote SSH URL address* used to access your Git backup repo
            * *Git remote SSH URL addresses* have the form `git@<gitserver>:<user>/<repo>.git`, which means that user account `<user>` has access to repository `<repo>` on Git server `<gitserver>`
            * To provide SSH access to your Git backup repo you have to add the generated public key to your `<user>` account on the `<gitserver>` server
                * See how to [set up an SSH key for BitBucket](https://confluence.atlassian.com/bitbucket/set-up-an-ssh-key-728138079.html)
                * See how to [connect to GitHub with SSH](https://help.github.com/articles/connecting-to-github-with-ssh/)
    * Run the following commands to deploy containers and see their logs (use `Ctrl-C` to exit)
```bash
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml pull
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml up -d
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml logs -f # to see the container logs in console; Ctrl-C to exit
```
    * If you didn't place the private key to the host persistent volume (as `${PERSISTENT_DIR}/dokuwiki/root/.ssh/id_rsa`), the container initialization script will generate a public/pravite key pair, store the generated keys in `${PERSISTENT_DIR}/dokuwiki/root/.ssh/`, and show the public key in the container log, waiting for the access to Git backup repo be provided - see the next point
    * If the container initialization script is not able to access Git backup repo, it will wait for 10 minutes (or till the moment the access is provided) checking once per minute for the access and asking you to add a public key. Look for the `Please add the public key ...` messages in the container log in console
    * On the first container run (after this installation procedure), fresh DokuWiki data will be `commited` to the configured Git backup repo. On the next container run, DokuWiki data from the Git backup repo will be `cloned/pulled` to the container `/data` volume
    * As script proceeds, point your browser to your wiki site URL to finish with DokuWiki installation wizard, fill in the form provided by the wizard, and click `Save`
    * The following message will appear in your browser
        * `The configuration was finished successfully. You may delete the install.php file now. ... `
    * Use `Ctrl-C` in console to exit from `docker-compose logs`, delete the `install.php` file with the following command:
        * `docker exec dokuwiki /bin/sh -c "rm -fr /var/www/install.php"`

## Next run ##

* If you did installation procedure before and just need to run existing 'dokuwiki' container, run the following command from the project directory (created during installation procedure):
```bash
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml pull
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml up -d
```

## Upgrade ##

* Use the the following commands to upgrade containers, it is recommended
```bash
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml down
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml pull
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml up -d
```
* You can also upgrade DokuWiki automatically through its UI

## Backup and Restore ##

* All data in `/data` folder are periodically backed up to the provided Git backup repo
* Any time you run a container from [this image](https://hub.docker.com/r/mtilson/dokuwiki/) on any host with configured access to Git backup repo, DokuWiki data from the repo will be synced with the container's `/data` volume and host's `${PERSISTENT_DIR}/dokuwiki/data` folder. Use `pre-deploy.sh` script and the above *Installation* section to prepare host

## License ##

* MIT. See `LICENSE` for more details


CHANGELOG.md
# Changelog

## 2020/01/15

* sepatate common `docker-compose.yml` to `docker-compose.yml` for 'dokuwiki' and `traefik/docker-compose.yml` for traefik

## 2019/04/15

* DokuWiki 2018-04-22b
* Alpine Linux 3.9

## 2018/05/16

* init based on `crazy-max/docker-dokuwiki`
* DokuWiki 2018-04-22a
* Alpine Linux 3.8
* PHP 7.2


deploy.sh
#!/usr/bin/env bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd -P)"
ME="$( basename "${BASH_SOURCE[0]}" )"
ENV="${DIR}/.env"

test -f "$ENV" || { echo "$ME: error: .env file ($ENV) doesn't exist, exiting" ; exit 255 ; }
source <(cat $ENV | grep "PERSISTENT_DIR\|GIT_BACKUP_REPO_URL")

# DOKUWIKI_FE_RULE
# PERSISTENT_DIR *
# ACME_EMAIL
# DOCKER_DOMAIN
# BACKUP_USER_EMAIL
# GIT_BACKUP_REPO_URL *
# TZ
# MEMORY_LIMIT
# UPLOAD_MAX_SIZE
# OPCACHE_MEM_SIZE

GIT_BACKUP_REPO_SERVER=$(echo $GIT_BACKUP_REPO_URL | cut -d"@" -f2 | cut -d":" -f1)
test -n "$GIT_BACKUP_REPO_SERVER" ||
    { echo "$ME: error: invalid git remote url ($GIT_BACKUP_REPO_URL) defined in .env file ($ENV), exiting" ; exit 255 ; }

test -n "${PERSISTENT_DIR}" ||
    { echo "$ME: error: persistent dir variable (PERSISTENT_DIR) is not defined in .env file ($ENV), exiting" ; exit 255 ; }

COMPOSE_FILE="docker-compose.yml"
test -n "$1" && COMPOSE_FILE="$1"

sudo rm -fr "${PERSISTENT_DIR}/dokuwiki"
sudo mkdir -p "${PERSISTENT_DIR}/dokuwiki"
echo "$ME: log: recreated ${PERSISTENT_DIR}/dokuwiki"

sudo touch "${PERSISTENT_DIR}/acme.json"
sudo chmod 600 "${PERSISTENT_DIR}/acme.json"
echo "$ME: log: refreshed ${PERSISTENT_DIR}/acme.json"

cd $DIR
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/docker-compose.yml > ${COMPOSE_FILE}

cat << _EOF
$ME: log: finished succesfully; run the following command to deploy containers and see their logs, use Ctrl-C to exit
docker-compose pull
docker-compose up -d
docker-compose logs -f
_EOF


entrypoint.sh
#!/bin/sh

test -n "$GIT_BACKUP_REPO_URL" ||
    { echo "$0: error: variable GIT_BACKUP_REPO_URL is undefined; exiting" ; exit 255 ; }

GIT_BACKUP_REPO_SERVER=$(echo $GIT_BACKUP_REPO_URL | cut -d"@" -f2 | cut -d":" -f1)
test -n "$GIT_BACKUP_REPO_SERVER" ||
    { echo "$0: error: invalid Git remote URL ($GIT_BACKUP_REPO_URL); exiting" ; exit 255 ; }

BACKUP_USER_EMAIL=${BACKUP_USER_EMAIL:-"dokuwiki-backup@example.com"}

TZ=${TZ:-"UTC"}
MEMORY_LIMIT=${MEMORY_LIMIT:-"256M"}
UPLOAD_MAX_SIZE=${UPLOAD_MAX_SIZE:-"16M"}
OPCACHE_MEM_SIZE=${OPCACHE_MEM_SIZE:-"128"}

function runas_nginx() {
  su - nginx -s /bin/sh -c "$1"
}

function plugins_and_templates() {
  echo -n "$0: log: bundled plugins saved: /tmp/bundledPlugins.txt, "
  bundledPlugins=$(ls -d /var/www/lib/plugins/*/ | cut -f6 -d'/')
  > /tmp/bundledPlugins.txt
  for bundledPlugin in ${bundledPlugins}; do
    echo "${bundledPlugin}" >> /tmp/bundledPlugins.txt
  done
  echo " $(wc -l < /tmp/bundledPlugins.txt) found"

  echo -n "$0: log: bundled templates saved: /tmp/bundledTpls.txt, "
  bundledTpls=$(ls -d /var/www/lib/tpl/*/ | cut -f6 -d'/')
  > /tmp/bundledTpls.txt
  for bundledTpl in ${bundledTpls}; do
    echo "${bundledTpl}" >> /tmp/bundledTpls.txt
  done
  echo " $(wc -l < /tmp/bundledTpls.txt) found"

  userPlugins=$(ls -l /data/plugins | egrep '^d' | awk '{print $9}')
  for userPlugin in ${userPlugins}; do
    if [ -d /var/www/lib/plugins/${userPlugin} ]; then
      echo "$0: log: WARNING: Plugin ${userPlugin} will not be used (already bundled in DokuWiki)"
      continue
    fi
    echo "$0: log: link: /var/www/lib/plugins/${userPlugin} -> /data/plugins/${userPlugin}"
    ln -sf /data/plugins/${userPlugin} /var/www/lib/plugins/${userPlugin}
    chown -h nginx: /var/www/lib/plugins/${userPlugin}
  done

  userTpls=$(ls -l /data/tpl | egrep '^d' | awk '{print $9}')
  for userTpl in ${userTpls}; do
    if [ -d /var/www/lib/tpl/${userTpl} ]; then
      echo "$0: log: WARNING: Template ${userTpl} will not be used (already bundled in DokuWiki)"
      continue
    fi
    echo "$0: log: link: /var/www/lib/tpl/${userTpl} -> /data/tpl/${userTpl}"
    ln -sf /data/tpl/${userTpl} /var/www/lib/tpl/${userTpl}
    chown -h nginx: /var/www/lib/tpl/${userTpl}
  done
}

echo "$0: log: timezone: ${TZ}"
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} > /etc/timezone

# PHP
echo "$0: log: php-fpm config file: /etc/php7/php-fpm.d/www.conf"
sed -e "s/@MEMORY_LIMIT@/$MEMORY_LIMIT/g" \
  -e "s/@UPLOAD_MAX_SIZE@/$UPLOAD_MAX_SIZE/g" \
  /tpls/etc/php7/php-fpm.d/www.conf > /etc/php7/php-fpm.d/www.conf

# OpCache
echo "$0: log: opcache config file: /etc/php7/conf.d/opcache.ini"
sed -e "s/@OPCACHE_MEM_SIZE@/$OPCACHE_MEM_SIZE/g" \
  /tpls/etc/php7/conf.d/opcache.ini > /etc/php7/conf.d/opcache.ini

# Nginx
echo "$0: log: nginx config file: /etc/nginx/nginx.conf"
sed -e "s/@UPLOAD_MAX_SIZE@/$UPLOAD_MAX_SIZE/g" \
  /tpls/etc/nginx/nginx.conf > /etc/nginx/nginx.conf

echo "$0: log: php preload file: /var/www/inc/preload.php"
cp -f /tpls/preload.php /var/www/inc/
chown nginx: /var/www/inc/preload.php

echo "$0: log: php install file: /var/www/install.php"
install_php_present=0
test -f /var/www/install.php && {
  sed -i "1s/.*/<?php define('DOKU_CONF', '\/data\/conf\/'); define('DOKU_LOCAL', '\/data\/conf\/');/" /var/www/install.php; 
  install_php_present=1;
  }

git config --global user.email "${BACKUP_USER_EMAIL}"
git config --global user.name "${BACKUP_USER_EMAIL%@*}"

cat > ~/.profile << EOF
alias ll='ls -la'
EOF

test -d ~/.ssh || {
  mkdir -p ~/.ssh
  chmod 700 ~/.ssh
  }

test -f ~/.ssh/config || {
  cat > ~/.ssh/config << EOF
Host $GIT_BACKUP_REPO_SERVER
    StrictHostKeyChecking no
EOF
  chmod 400 ~/.ssh/config
  }

test -f ~/.ssh/id_rsa || {
  ssh-keygen -t rsa -N "" -f ~/.ssh/id_rsa -C "${BACKUP_USER_EMAIL}"
  }
chmod 600 ~/.ssh/id_rsa

message=">>\n>> Please add the public key corresponding to the private one from ~/.ssh/id_rsa to your $GIT_BACKUP_REPO_SERVER git server account\n>>\n"
test -f ~/.ssh/id_rsa.pub && { 
  echo "Public key for $GIT_BACKUP_REPO_SERVER git server for backup:"
  cat ~/.ssh/id_rsa.pub
  message=">>\n>> Please add the public key shown above to your $GIT_BACKUP_REPO_SERVER git server account\n>>\n"
  }

count=0
while true
do
    git ls-remote "$GIT_BACKUP_REPO_URL" &>-
    if [ "$?" -eq 0 ]; then
        echo "$0: log: access to $GIT_BACKUP_REPO_URL is available"
        break
    else
        let "count++"
        delay=60

        echo "$0: log: access to $GIT_BACKUP_REPO_URL is not available"
        echo -e "$message"
        echo "$0: log: sleeping for $delay seconds."
        sleep $delay

        test $count -ne 10 || { echo "$0: error: failed to get access to $GIT_BACKUP_REPO_URL for $count times, exiting"; exit 255; }
    fi
done

data_commited=0
if [ ! -d /data/.git ]; then
    echo "$0: log: clone $GIT_BACKUP_REPO_URL to /data"
    rm -fr /data/*
    git clone $GIT_BACKUP_REPO_URL /data

    test -d /data/.git/objects || { echo "$0: error: no /data/.git/objects after cloning repo, exiting"; exit 255; }
    object_count=$(find /data/.git/objects -type f | wc -l)

    test -n "$object_count" || { echo "$0: error: number of objects is empty, exiting"; exit 255; }
    if [ "$object_count" -eq "0" ]; then

        mkdir -p /data/plugins /data/tpl
        touch /data/plugins/.dummy
        touch /data/tpl/.dummy
        cp -Rf /var/www/conf /data/
        cp -Rf /var/www/data /data/
        cp -f /tpls/local.protected.php /data/conf
        cp -f /tpls/.gitignore /data/

        plugins_and_templates

        chown -R nginx: /data

        echo "$0: log: commit /data to $GIT_BACKUP_REPO_URL" 
        cd /data
        git add -A
        git commit -m "wiki created @ `date -u`"
        git push -u origin master

        data_commited=1
    else
        echo "$0: log: cloned $GIT_BACKUP_REPO_URL to /data"

        test -d /data/data/cache || mkdir /data/data/cache # all 4 dirs are ignored by git in .gitignore
        test -d /data/data/index || mkdir /data/data/index
        test -d /data/data/locks || mkdir /data/data/locks
        test -d /data/data/tmp || mkdir /data/data/tmp

        chown -R nginx: /data
    fi
else
    echo "$0: log: pull $GIT_BACKUP_REPO_URL to /data"
    cd /data
    git pull origin master
    chown -R nginx: /data
fi

if [ $install_php_present -eq "1" -a $data_commited -eq "0" ]; then
  echo "$0: log: /var/www/install.php present - container was initially deployed or redeployed"
  echo "$0: log: /data was not committed yet - as either 1) local repo was not present, it was cloned, but it is not empty; or 2) local repo existed and was pulled"
  echo "$0: log: we need to apply and commit configuration changes of possible DokuWiki project upgrade"
  cp -Rf /var/www/conf /data/
  plugins_and_templates
  chown -R nginx: /data

  cd /data
  git add -A
  git commit -m "wiki updated @ `date -u`"
  git push -u origin master
fi

if [ -f /data/conf/local.php -o -f /data/conf/users.auth.php -o -f /data/conf/acl.auth.php ]; then
  rm -fr /var/www/install.php
  echo "$0: log: launching '/var/www/bin/indexer.php -c'"
  runas_nginx 'php7 /var/www/bin/indexer.php -c'
else
  echo -e ">>\n>> Please, point your browser to DokuWiki installation wizard page (/install.php) of your wiki site to finish installation\n>>\n"
fi

exec "$@"


pre-deploy.sh
#!/usr/bin/env bash

ME="$( basename "${BASH_SOURCE[0]}" )" # script short name
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd -P)" # script dir name

test -e "${DIR}/.env" && source <(cat "${DIR}/.env" | grep "^PERSISTENT_DIR=\|^GIT_BACKUP_REPO_URL=")
test -e "${PWD}/.env" && source <(cat "${PWD}/.env" | grep "^PERSISTENT_DIR=\|^GIT_BACKUP_REPO_URL=")
PERSISTENT_DIR=${PERSISTENT_DIR:-/opt/docker/persistent}

test -n "${GIT_BACKUP_REPO_URL}" || 
    { echo "$ME: error: variable GIT_BACKUP_REPO_URL is undefined (use '.env' file in the repo dir and/or in the script working dir to assign Git remote URL to GIT_BACKUP_REPO_URL variable); exiting" ;
      exit 255 ;
    }

GIT_BACKUP_REPO_SERVER=$(echo $GIT_BACKUP_REPO_URL | cut -d"@" -f2 | cut -d":" -f1)
test -n "$GIT_BACKUP_REPO_SERVER" ||
    { echo "$ME: error: invalid Git remote URL ($GIT_BACKUP_REPO_URL) defined using '.env' file in the repo dir and/or the script working dir; exiting" ;
      exit 255 ;
    }

sudo rm -fr "${PERSISTENT_DIR}/dokuwiki"
sudo mkdir -p "${PERSISTENT_DIR}/dokuwiki/root/.ssh"
sudo touch "${PERSISTENT_DIR}/acme.json"
sudo chmod 600 "${PERSISTENT_DIR}/acme.json"

(
cd $DIR
mkdir -p traefik
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/docker-compose.yml > docker-compose.yml
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/traefik/docker-compose.yml > traefik/docker-compose.yml
)

echo "We are ready. Go to the repo dir ($DIR) and use 'docker-compose -f docker-compose.yml -f traefik/docker-compose.yml CMD' command to proceed."


deploy-dokuwiki.sh
#!/usr/bin/env bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd -P)"
ME="$( basename "${BASH_SOURCE[0]}" )"
ENV="${DIR}/.env"

test -f "$ENV" || { echo "$ME: error: .env file ($ENV) doesn't exist, exiting" ; exit 255 ; }
source <(cat $ENV | grep "PERSISTENT_DIR\|GIT_BACKUP_REPO_URL")

# DOKUWIKI_FE_RULE
# PERSISTENT_DIR *
# ACME_EMAIL
# DOCKER_DOMAIN
# BACKUP_USER_EMAIL
# GIT_BACKUP_REPO_URL *
# COMMON_NETWORK
# TZ
# MEMORY_LIMIT
# UPLOAD_MAX_SIZE
# OPCACHE_MEM_SIZE

GIT_BACKUP_REPO_SERVER=$(echo $GIT_BACKUP_REPO_URL | cut -d"@" -f2 | cut -d":" -f1)
test -n "$GIT_BACKUP_REPO_SERVER" ||
    { echo "$ME: error: invalid git remote url ($GIT_BACKUP_REPO_URL) defined in .env file ($ENV), exiting" ; exit 255 ; }

test -n "${PERSISTENT_DIR}" ||
    { echo "$ME: error: persistent dir variable (PERSISTENT_DIR) is not defined in .env file ($ENV), exiting" ; exit 255 ; }

sudo rm -fr "${PERSISTENT_DIR}/dokuwiki"
sudo mkdir -p "${PERSISTENT_DIR}/dokuwiki"
sudo mkdir -p "${PERSISTENT_DIR}/dokuwiki/root/.ssh"
echo "$ME: log: recreated ${PERSISTENT_DIR}/dokuwiki"

sudo touch "${PERSISTENT_DIR}/acme.json"
sudo chmod 600 "${PERSISTENT_DIR}/acme.json"
echo "$ME: log: refreshed ${PERSISTENT_DIR}/acme.json"

cd $DIR
mkdir -p traefik
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/docker-compose.yml > docker-compose.yml
curl -sSL https://raw.githubusercontent.com/mtilson/dokuwiki/master/traefik/docker-compose.yml > traefik/docker-compose.yml

cat << _EOF
$ME: log: finished succesfully; run the following command to deploy containers and see their logs, use Ctrl-C to exit
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml pull
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml up -d
docker-compose -f docker-compose.yml -f traefik/docker-compose.yml logs -f
_EOF
